web/public/javascripts/routers/datasets.js

```diff
@@ -201,43 +200,62 @@ App.DatasetsRoute = Ember.Route.extend({
 });
 
 App.DatasetRoute = Ember.Route.extend({
-  setupController: function(controller, params) {
-    var _this = this;
-    detailController = controller;
+  setupController(controller, params) {
     currentTab = 'Datasets';
     updateActiveTab();
     var id = 0;
     var source = '';
     var urn = '';
     var name = '';
-    controller.set("hasProperty", false);
-    if(params && params.id)
-      {
-        id = params.id;
-        source = params.source;
-        urn = params.urn;
-        name = params.name;
-        datasetController.set("detailview", true);
-        if (params.originalSchema)
-        {
-          Ember.set(params, 'schema', params.originalSchema);
+
+    /**
+     * Parses a JSON dataset schema representation and extracts the field names and types into a list of maps
+     * @param {JSON} schema
+     * @returns {Array.<*>}
+     */
+    const getFieldNamesAndTypesFrom = schema => {
+      const getFieldTypeSet = ({fields} = {fields: []}) => fields.map(({name, type: [firstType, ...type], fields}) => {
+        if (fields) {
+          return getFieldTypeSet({fields});
         }
-        controller.set('model', params);
 
+        return {
+          name,
+          type: firstType === 'null' ? type : [firstType, ...type]
+        };
+      });
+
+      // Flatten nested structure if present
+      return [].concat(...getFieldTypeSet(JSON.parse(schema))); // TODO: cover n-th dimension, if expected
+    };
+    controller.set("hasProperty", false);
+
+
+    if (params && params.id) {
+      ({id, source, urn, name} = params);
+      let originalSchema = params;
+
+      datasetController.set('detailview', true);
+      if (originalSchema) {
+        Ember.set(params, 'schema', originalSchema);
       }
-      else {
-        if (params.dataset)
-          {
-            id = params.dataset.id;
-            source = params.dataset.source;
-            urn = params.dataset.urn;
-            name = params.dataset.name;
-            controller.set('model', params.dataset);
-            datasetController.set("detailview", true);
-          }
-      }
 
-      var instanceUrl = 'api/v1/datasets/' + id + "/instances";
+      controller.set('model', params);
+    } else if (params.dataset) {
+      ({dataset: {id, source, urn, name}} = params);
+
+      controller.set('model', params.dataset);
+      datasetController.set('detailview', true);
+    }
+
+    controller.set('datasetId', id);
+    if (params.dataset.schema || params.schema) {
+      let {schema} = params.dataset || params;
+      controller.set('datasetSchemaFieldsAndTypes', getFieldNamesAndTypesFrom(schema));
+      controller.set('securitySpec', params.securitySpec);
+    }
+
+    var instanceUrl = 'api/v1/datasets/' + id + "/instances";
```

Comment:
Sure, this was actually old code. My change here was line formatting, that changed the indentation for line 258.